File: code/1022/1022.rb
number_of_lines = gets.to_i

def mdc(num1, num2)
  return num1 if num2.zero?

  mdc(num2, num1 % num2)
end

def sum_fractions(fraction1, fraction2)
  denominator = fraction1[1] * fraction2[1]
  numerator =
    (denominator / fraction1[1]) * fraction1[0] +
    (denominator / fraction2[1]) * fraction2[0]

  [numerator, denominator]
end

def sub_fractions(fraction1, fraction2)
  denominator = fraction1[1] * fraction2[1]
  numerator =
    (denominator / fraction1[1]) * fraction1[0] -
    (denominator / fraction2[1]) * fraction2[0]

  [numerator, denominator]
end

def mul_fractions(fraction1, fraction2)
  denominator = fraction1[1] * fraction2[1]
  numerator = fraction1[0] * fraction2[0]

  [numerator, denominator]
end

def div_fractions(fraction1, fraction2)
  denominator = fraction1[1] * fraction2[0]
  numerator = fraction1[0] * fraction2[1]

  [numerator, denominator]
end

def simplify_fractions(num1, num2)
  div = mdc(num1.abs, num2)
  [num1 / div, num2 / div]
end

def print_fractions(fraction, simplify_fraction)
  puts "#{fraction[0]}/#{fraction[1]} = #{simplify_fraction[0]}/#{simplify_fraction[1]}"
end

[*1..number_of_lines].each do
  entry = gets.split(' ')
  fraction = []

  if entry[3] == '+'
    fraction = sum_fractions(
      [entry[0].to_i, entry[2].to_i],
      [entry[4].to_i, entry[6].to_i]
    )
  end
  if entry[3] == '-'
    fraction = sub_fractions(
      [entry[0].to_i, entry[2].to_i],
      [entry[4].to_i, entry[6].to_i]
    )
  end
  if entry[3] == '/'
    fraction = div_fractions(
      [entry[0].to_i, entry[2].to_i],
      [entry[4].to_i, entry[6].to_i]
    )
  end
  if entry[3] == '*'
    fraction = mul_fractions(
      [entry[0].to_i, entry[2].to_i],
      [entry[4].to_i, entry[6].to_i]
    )
  end

  simplified_fraction = simplify_fractions(fraction[0], fraction[1])

  print_fractions(fraction, simplified_fraction)
end

File: code/1022/1022.js
const input = require("fs").readFileSync("/dev/stdin", "utf8");
const lines = input.split("\n");

const mdc = (a, b) => {
  if (b === 0) return a;
  return mdc(b, a % b);
};

const sum2fraction = (fraction1, fraction2) => {
  const denominator = fraction1[1] * fraction2[1];
  const numerator =
    (denominator / fraction1[1]) * fraction1[0] +
    (denominator / fraction2[1]) * fraction2[0];

  return [numerator, denominator];
};

const sub2fraction = (fraction1, fraction2) => {
  const denominator = fraction1[1] * fraction2[1];
  const numerator =
    (denominator / fraction1[1]) * fraction1[0] -
    (denominator / fraction2[1]) * fraction2[0];

  return [numerator, denominator];
};

const mul2fraction = (fraction1, fraction2) => {
  const denominator = fraction1[1] * fraction2[1];
  const numerator = fraction1[0] * fraction2[0];

  return [numerator, denominator];
};

const div2fraction = (fraction1, fraction2) => {
  const denominator = fraction1[1] * fraction2[0];
  const numerator = fraction1[0] * fraction2[1];

  return [numerator, denominator];
};

const simplifyFraction = (a, b) => {
  const div = mdc(Math.abs(a), b);
  return [a / div, b / div];
};

const printFractions = (fraction, simplifyFraction) => {
  console.log(
    `${fraction[0]}/${fraction[1]} = ${simplifyFraction[0]}/${simplifyFraction[1]}`
  );
};

for (let i = 1; i < lines.length - 1; i++) {
  const entry = lines[i].split(" ");
  let fraction;

  if (entry[3] === "+") {
    fraction = sum2fraction(
      [parseInt(entry[0]), parseInt(entry[2])],
      [parseInt(entry[4]), parseInt(entry[6])]
    );
  }
  if (entry[3] === "-") {
    fraction = sub2fraction(
      [parseInt(entry[0]), parseInt(entry[2])],
      [parseInt(entry[4]), parseInt(entry[6])]
    );
  }
  if (entry[3] === "/") {
    fraction = div2fraction(
      [parseInt(entry[0]), parseInt(entry[2])],
      [parseInt(entry[4]), parseInt(entry[6])]
    );
  }
  if (entry[3] === "*") {
    fraction = mul2fraction(
      [parseInt(entry[0]), parseInt(entry[2])],
      [parseInt(entry[4]), parseInt(entry[6])]
    );
  }

  const simplifiedFraction = simplifyFraction(fraction[0], fraction[1]);

  printFractions(fraction, simplifiedFraction);
}
